### PR TITLE
Add new vote state version that replaces Lockout with LandedVote to a…

### PR DESCRIPTION
**Pull request by bji**
…llow vote latency to be tracked in a future change.

Includes a feature to be enabled which will when enabled cause the vote state to be written in the new form.

### CODE REVIEW

**Comment by bji** [comment]
This pull request is identical to the previous pull request for this feature, but with the following additions:

1. ashwinsekar's fix to the mismatch hash issue has been included.  The underlying issue was that the upgrade function which took the 1.14.11 version of VoteState and transformed it into the current version, was setting the prior_voters field to a default value instead of copying the value from the prior state.  This lost data from the vote account which then meant that the vote account hash no longer lined up with what the rest of the cluster computed.  This was an oversight on my part; when I made the upgrade function somehow I just didn't set the value correctly.

2. ashwinsekar made a change to ensure that rent exempt minimum and evaluations of expected vote account state size are feature-gated, whereas previously only the "new" values were being used for those purposes.  This mattered in a few places:

- The vote program's "InitalizeAccount" instruction handling was checking to make sure that the size of the account being initialized as a VoteAccount matched the "new" size instead of the feature-gated vote account size.  Unless the feature-gated size is used here, new software diverges from old software in the behavior of the vote program (different versions failing on different account sizes).

- The StakeCache's "check_and_store" function in the new code would evict all vote accounts since their size didn't match the "new size".  Instead, the check needs to be against the feature-gated size.

- Similarly the StakeCache's "Stakes" struct constructor would reject vote accounts because of the size mismatch.

3. ashwinsekar made a change to fix the hard-coded offset of the "prior voters" field within the vote state struct.  I failed to update this value in my initial change.  Having the wrong value contributed to cases of failure in the StakeCache mentioned previously.

4. I improved the test_is_correct_size_and_initialized() test to ensure that it covers the new cases.  I verified that it failed before the fixes and succeeds after the fixes.

5. I improved the vote_state_upgrade_from_1_14_11() test to ensure that it covers the prior_voters field, which unfortunately it did not, which is what allowed the issue originally causing point (1) to occur.  I verified that the test failed before the fix for (1) was applied, and succeeded after.

**Comment by AshwinSekar** [comment]
Currently running this against mnb (non voting, non staked). `9bszuZP8FkD6YuNikRQ5MnvZmAxNr79bA8k4E9ic18k1`. Will monitor to see if any issues pop up.

**Review by AshwinSekar — APPROVED** [review]
No issues found when running against mnb, think we caught everything with the patch.

File: sdk/program/src/vote/state/mod.rs [hunk 1]
```diff
@@ -302,7 +338,7 @@ impl VoteState {
     /// Upper limit on the size of the Vote State
     /// when votes.len() is MAX_LOCKOUT_HISTORY.
     pub const fn size_of() -> usize {
-        3731 // see test_vote_state_size_of.
+        3762 // see test_vote_state_size_of.
```
Comment (steveluscher) [inline sdk/program/src/vote/state/mod.rs:341]: This broke the JavaScript client's ability to create new accounts. See https://github.com/solana-labs/solana-web3.js/pull/1306

**Comment by steveluscher** [comment]
What would you think about making this less restrictive, in the sense that it could pass if the account size was _at least_ as large as necessary.

https://github.com/solana-labs/solana/blob/174951478f11719c96483777f7eb6030eb6303ef/programs/vote/src/vote_state/mod.rs#L973-L976

That would let us update clients (eg. Rust, JS) _now_ in a way that's backward compatible (ie. supplying an account larger than necessary would not fatal old clients, and would be perfectly sized for new ones).

**Comment by t-nelson** [comment]
> What would you think about making this less restrictive, in the sense that it could pass if the account size was _at least_ as large as necessary.
> 
> https://github.com/solana-labs/solana/blob/174951478f11719c96483777f7eb6030eb6303ef/programs/vote/src/vote_state/mod.rs#L973-L976
> 
> That would let us update clients (eg. Rust, JS) _now_ in a way that's backward compatible (ie. supplying an account larger than necessary would not fatal old clients, and would be perfectly sized for new ones).

prolly a bad idea. vote accounts are about the last account we want people overallocating for the lulz

**Comment by steveluscher** [comment]
Wait, this isn't all for the lulz?

**Comment by steveluscher** [comment]
Sounds good. I subscribed to the feature activation issue and put a note in there to ship the client change when we flip the switch.

**Comment by t-nelson** [comment]
> Wait, this isn't all for the lulz?

well yeah, but this particular lulz might disturb my beauty rest

**Comment by AshwinSekar** [comment]
can the client check the feature set before sending the tx? that's what we're doing for local cluster atm 
https://github.com/solana-labs/solana/blob/05c932f92d3a23c795d689ea2e19d50bf5fbde48/local-cluster/src/local_cluster.rs#LL663C1-L686C1

**Comment by steveluscher** [comment]
> can the client check the feature set…

When we're talking about web3.js we're typically talking about _web_ clients. Web clients, being consumer applications, are incredibly performance sensitive, in the sense that customers will rage if their website loads in more than 1.5 seconds.

It's out of the question to add round trips just to be able to boot up a website.

1. Ask for website.
2. Web server responds with HTML
3. Browser goes ‘oh shit I need to download some JavaScript to even know what I am.’
4. Fetches JavaScript
5. Parse/JIT/execute
6. Hey, I'm web3.js! But I don't know how to behave, better go fetch some feature flags from an RPC
7. Get feature flags from RPC
8. Oh, this is the stuff I support – cool time to do website stuff

By the time you hit step 6 your customer has already closed the window and opened TikTok.

**Comment by steveluscher** [comment]
I know that ‘initializing vote accounts’ is not really something that you'll find on the startup path for a website, and we _could_ lazy load the feature set.

I also know that the situation is nuanced, and not all client code runs in an environment where it can contact random servers in deterministic ways (eg. server-side rendering).

**Pull request merged**